husbandry.lua

```diff
@@ -0,0 +1,326 @@
+    local job = ic.make_job()
+    job.job_type = df.job_type.MilkCreature
+    dfhack.job.addGeneralRef(job, df.general_ref_type.UNIT_MILKEE, unit.id)
+    ic.assignToWorkshop(job, workshop)
+end
+
+
+-- configuration management
+
+GLOBAL_KEY = 'husbandry'
+
+local function get_default_state()
+    return {
+        enabled = false,
+        milking = true,
+        shearing = true,
+        roaming = true;
+        pasture = false
+    }
+end
+
+state = state or get_default_state()
+
+function isEnabled()
+    return state.enabled
+end
+
+function persist_state()
+    dfhack.persistent.saveSiteData(GLOBAL_KEY, {
+        enabled=state.enabled,
+        milking=state.milking,
+        shearing=state.shearing,
+        roaming=state.roaming,
+        pasture=state.pasture,
+    })
+end
+
+--- Load the saved state of the script
+local function load_state()
+    -- load persistent data
+    local persisted_data = dfhack.persistent.getSiteData(GLOBAL_KEY, get_default_state())
+    state.enabled = persisted_data.enabled
+    state.milking = persisted_data.milking
+    state.shearing = persisted_data.shearing
+    state.roaming = persisted_data.roaming
+    state.pasture = persisted_data.pasture
+    return state
+end
+
+-- main script action
+
+local function action()
+    debug('husbandry: running loop')
+
+    -- organize workshops by allowed labors and z-level
+    ---@type table<integer,df.building_workshopst[]>
+    local farmer_shearing = {}
+    ---@type table<integer,df.building_workshopst[]>
+    local farmer_milking = {}
+    for _, workshop in ipairs(df.global.world.buildings.other.WORKSHOP_FARMER) do
+        if not workshop.profile.blocked_labors[df.unit_labor.SHEARER] then
+            table.insert(ensure_key(farmer_shearing, workshop.z), workshop)
+        end
+        if not workshop.profile.blocked_labors[df.unit_labor.MILK] then
+            table.insert(ensure_key(farmer_milking, workshop.z), workshop)
+        end
+    end
+
+    -- gather units that are already being milked or sheared
+    ---@type table<integer,boolean>
+    local unit_milking = {}
+    ---@type table<integer,boolean>
+    local unit_shearing = {}
+
+    -- go over all workshops to to catch player-initiated jobs
+    for _, workshop in ipairs(df.global.world.buildings.other.WORKSHOP_FARMER) do
+        for _, job in ipairs(workshop.jobs) do
+            if state.milking and job.job_type == df.job_type.MilkCreature then
+                local milkee = dfhack.job.getGeneralRef(job, df.general_ref_type.UNIT_MILKEE)
+                if milkee then
+                    unit_milking[milkee.unit_id] = true
+                end
+            elseif state.shearing and job.job_type == df.job_type.ShearCreature then
+                local shearee  = dfhack.job.getGeneralRef(job, df.general_ref_type.UNIT_SHEAREE)
+                if shearee then
+                    unit_shearing[shearee.unit_id] = true
+                end
+            end
```

Comment:
can't help but notice that this code is somewhat repetitive

also this pattern appears likely to recur, may be worth abstracting to a library routine